### PR TITLE
[SPARK-24085][SQL] Query returns UnsupportedOperationException when scalar subquery is present in partitioning expression

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
In this case, the partition pruning happens before the planning phase of scalar subquery expressions.
For scalar subquery expressions, the planning occurs late in the cycle (after the physical planning)  in "PlanSubqueries" just before execution. Currently we try to execute the scalar subquery expression as part of partition pruning and fail as it implements Unevaluable.

The fix attempts to ignore the Subquery expressions from partition pruning computation. Another option can be to somehow plan the subqueries before the partition pruning. Since this may not be a commonly occuring expression, i am opting for a simpler fix.

Repro
``` SQL
CREATE TABLE test_prc_bug (
id_value string
)
partitioned by (id_type string)
location '/tmp/test_prc_bug'
stored as parquet;

insert into test_prc_bug values ('1','a');
insert into test_prc_bug values ('2','a');
insert into test_prc_bug values ('3','b');
insert into test_prc_bug values ('4','b');


select * from test_prc_bug
where id_type = (select 'b');
```
## How was this patch tested?
Added test in SubquerySuite and hive/SQLQuerySuite

### CODE REVIEW

**Comment by maropu** [comment]
Why the orc works fine?

**Comment by dilipbiswal** [comment]
@maropu Thanks for your response. ORC has CONVERT_METASTORE_ORC set to false as default. So its not converted to a file based datasource. If we set this to true then we would see the same issue for ORC. I have added test to cover the case.

**Comment by maropu** [comment]
Ah, ok. Thanks!

**Comment by maropu** [comment]
plz add `[SQL]` in the title.

**Comment by maropu** [comment]
One question; we have no risk to miss any partition pruning by this fix?

**Comment by dilipbiswal** [comment]
@maropu So with the fix, if the query predicate contains an scalar subquery expression, then that expression is not considered for partition pruning.  For example, if the predicate was , part_key1 = (select ...) and part_key2 = 5 , then only the 2nd part of the expression is considered for pruning purposes and the first part will be a regular filter.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 1]
```diff
@@ -955,4 +955,28 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
...
+    withTempPath { tempDir =>
+      withTable("parquet_part") {
+        sql(
+          s"""
+             |CREATE TABLE parquet_part (id_value string, id_type string)
+             |USING PARQUET
+             |OPTIONS (
+             |  path '${tempDir.toURI}'
+             |)
+             |PARTITIONED BY (id_type)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]:
```Scala
        Seq("1" -> "a", "2" -> "a", "3" -> "b", "4" -> "b")
          .toDF("id_value", "id_type")
          .write
          .mode(SaveMode.Overwrite)
          .partitionBy("id_type")
          .format("parquet")
          .saveAsTable("parquet_part")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala [hunk 2]
```diff
@@ -55,7 +55,9 @@ private[sql] object PruneFileSourcePartitions extends Rule[LogicalPlan] {
           partitionSchema, sparkSession.sessionState.analyzer.resolver)
       val partitionSet = AttributeSet(partitionColumns)
       val partitionKeyFilters =
-        ExpressionSet(normalizedFilters.filter(_.references.subsetOf(partitionSet)))
+        ExpressionSet(normalizedFilters.
+          filterNot(SubqueryExpression.hasSubquery(_)).
+          filter(_.references.subsetOf(partitionSet)))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]:
```Scala
        ExpressionSet(normalizedFilters
          .filterNot(SubqueryExpression.hasSubquery)
          .filter(_.references.subsetOf(partitionSet)))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 3]
```diff
@@ -76,7 +76,10 @@ object FileSourceStrategy extends Strategy with Logging {
           fsRelation.partitionSchema, fsRelation.sparkSession.sessionState.analyzer.resolver)
       val partitionSet = AttributeSet(partitionColumns)
       val partitionKeyFilters =
-        ExpressionSet(normalizedFilters.filter(_.references.subsetOf(partitionSet)))
+        ExpressionSet(normalizedFilters.
+          filterNot(SubqueryExpression.hasSubquery(_)).
+          filter(_.references.subsetOf(partitionSet)))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: The same here

**Comment by gatorsmile** [comment]
This fix is safe for us to backport to the previous versions. To achieve a better performance, we can start a separate job to execute these subqueries. : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: @gatorsmile Will make the change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: @gatorsmile Will make the change.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala (hunk 1 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: @gatorsmile OK.

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks a lot. Addressed the comments.

**Comment by gatorsmile** [comment]
LGTM

**Comment by dilipbiswal** [comment]
@gatorsmile @maropu Thank you very much !!

**Pull request closed without merging**